### ISSUE TITLE
Investigate CI solutions

### ISSUE THREAD

**Issue by StephenHodgson** [Microsoft/MixedRealityToolkit-Unity#457]
Hi Everyone,
I'm interested in setting up an continuous integration server to automatically test and make releases.

Does anyone have any suggestions?
I was looking at free to use solutions like [Travis-CI](https://travis-ci.org/) or [Appveyor](https://www.appveyor.com/).

The main features I think we should look for is:

- GitHub Releases
- Build Testing on Pull Request Merge

Does anyone have experience doing this?

(related to https://github.com/Microsoft/HoloToolkit-Unity/issues/217)

**Comment by HodgsonSDAS** [comment]
I found a nice blog entry about [utilizing GitHub, Unity3D, and Visual Studio Team Services](http://afuzzyllama.github.io/2016/02/25/github-unity3d-and-visual-studio-team-services.html)

**Comment by aalmada** [comment]
I've been using Appveyor for non-Unity projects. These is a legal issue with the Unity libraries required to build: http://help.appveyor.com/discussions/suggestions/1028-support-for-unity-3d

**Comment by HodgsonSDAS** [comment]
https://youtu.be/kSXomLkMR68?t=24m45s

**Comment by jessemcculloch** [comment]
What is @riverar using to create the unity packages for his site?  Something that can be leveraged, or not?

**Comment by HodgsonSDAS** [comment]
Great question!

I was hoping to leverage his setup but I think MS is kinda frowning upon that because they don't have control of it.

**Comment by NeerajW** [comment]
Clarifying @HodgsonSDAS's statement to make intent more clear :)

I am not in favor of hosting packages on external websites which are not manageable easily using OSS.
I would recommend hosting packages in the Externals or UnityPackages folder in the same repository.
Would be happy to have riverar's updates to this repository if we deem that better.
Con:  Adding Unity packages increases repo size
Pro: Having unity packages makes versioning so much simpler for people.

**Comment by HodgsonSDAS** [comment]
http://www.infoworld.com/article/3171380/cloud-computing/visual-studios-cloud-ide-gets-smarter-about-builds.html

**Comment by Zod-** [comment]
Unity cloud builds would be interesting if they finally added UWP builds to it. I did some research today and got it integrated into our VSTS. I even got the test results migrated from the Unity build artifacts to display in VSTS.
![ztrxuys](https://user-images.githubusercontent.com/5775094/30673243-f1bd6236-9e71-11e7-8e3d-3cfa8d4117bd.png)
```csharp
        public void SimpleTest()
        {
            #if !FAIL_TEST
            Assert.That(true, Is.True);
            #else
            Assert.That(false, Is.True);
            #endif
        }
```
I still have to make a proper task for the whole process but I got the general flow down.

**Comment by riverar** [comment]
Just saw this thread, sorry. @jessemcculloch @StephenHodgson

I stopped my site thinking this repo was already producing CI builds. Do I need to spin it back up?

To answer the question: You'll want to build a pipeline that cleans up the repo (e.g. delete examples, meta folders if desired), then call out to Unity. For example, I used: `"C:\Program Files\Unity\Editor\Unity.exe" -batchmode -exportPackage Assets "%BUILD_COLLECTION_ROOT%\HoloToolkit-%BUILD_CURRENT_SHORT_SHA%.unitypackage" -projectPath C:\Build\HoloToolkit-Unity -quit`. Beware Unity (incorrectly) tries to bind to OpenGL so you'll need some filler libs if using Server Core or Nano.

I also scripted some azcopy to upload automagically.

Was a simple batch file on a tiny Server Core instance on my home server. 🤣
